OSX Support

danwalmsley opened the pull request:
Adds support for OSX.
- Also updates code to support latest version of Avalonia

nopara73 commented:
# Testing On Windows

## Issue 1

After `dotnet run` there's a warning:

![image](https://user-images.githubusercontent.com/9156103/46696200-e4352a80-cc11-11e8-8497-a9c08dd0702f.png)

## Issue 2

Cursored select highlighting is a new feature, which is nice, but the contrast is wrong:

![image](https://user-images.githubusercontent.com/9156103/46696347-3e35f000-cc12-11e8-8d1d-6c2cf3de6af0.png)

This issue is happening universally.

## Issue 3

Something's wrong on the History tab.

![image](https://user-images.githubusercontent.com/9156103/46696751-375bad00-cc13-11e8-9019-67a14b312928.png)

## Issue 4

Highlighting is too much here:

![image](https://user-images.githubusercontent.com/9156103/46696903-a507d900-cc13-11e8-9a11-b4fd803f48bb.png)

## Issue 5

More highlighting issues.

![image](https://user-images.githubusercontent.com/9156103/46697006-e9937480-cc13-11e8-9ac5-8e5ea7c2379a.png)

## Issue 6

Legal documents have glitches like this.

![image](https://user-images.githubusercontent.com/9156103/46697160-51e25600-cc14-11e8-9a7c-6f3b4e8a5392.png)

Zolgarr commented:
### macOS High Sierra - Review

**Issue** 1

In general it's quite slow (e.g. if I bring the cursor somewhere, the highlighting is a bit slow, if I click on something, it's a bit slow, etc.)

**Issue** 2

The highlighting is not really visible.

<img width="329" alt="highlighting" src="https://user-images.githubusercontent.com/42039790/46812857-e3b5a480-cd75-11e8-977d-3c76fc47eb15.png">

**Issue** 3 

I opened Wasabi Wallet, and it didn't bring the original home screen, just the upper bar (it happened only once though)

<img width="1440" alt="nothing" src="https://user-images.githubusercontent.com/42039790/46812979-24152280-cd76-11e8-9469-c461d748c865.png">

**Issue** 4 

If I click on maximize, it's not actually on full screen. This is a bit frustrating if I want to swipe to home screen, it's showing Wasabi Wallet then. On the picture attached, you can see that there's no option to do full screen on the mac upper bar. 

<img width="338" alt="no fullscreen" src="https://user-images.githubusercontent.com/42039790/46813290-bf0dfc80-cd76-11e8-9e55-61e49ece26f5.png">

**Issue** 5 

If it's minimized, the label is still there.

<img width="1440" alt="minimize" src="https://user-images.githubusercontent.com/42039790/46813452-03010180-cd77-11e8-876a-b6e934713a49.png">

**Issue** 6 

If I close Wasabi Wallet, there's a message that pops up. It says that dotnet has unexpectedly quit. I can reopen again, leave it, or send a report which will be sent to Apple.

<img width="1440" alt="quit" src="https://user-images.githubusercontent.com/42039790/46813484-1318e100-cd77-11e8-82ef-0c17af64398f.png">

**Issue** 7

I closed Wasabi Wallet and it popped back up again.

**Issue** 8 

It doesn't react to certain commands. I click on minimize or on restore, and it doesn't do anything.

danwalmsley commented:
@nopara73 
Issue 1 - avalonia warnings fixed

Issue 2,4,5 basically the same cause and should be fixed, check if you approve of the highlighting now.

Issue 3, fixed

Issue 6, fixed by @lontivero

danwalmsley commented:
@Zolgarr 
Issue 1 - aware, we are implementing GPU rendering and something called DisplayLink to fix this, will take another week probably.

Issue 2 - should be fixed now

Issue 3 - I have also seen this, may be related to rendering code, so we will see if this goes away once we upgrade renderer.

Issue 4 - will look into.

Issue 5 - will look into (this could be a bug affecting windows and Linux also?)

Issue 6 - fixed

Zolgarr commented:
### @danwalmsley  

Regarding the issues above:

**Issue 2**
The highlighting is great! 

**Issue 4 (and also Issue 8)**
It works! Awesome!

**Issue 6**
It worked well once, but at other times it popped that message up.

**Issue 7**
By this I meant that I closed Wasabi Wallet, and then it popped back up as if I reopened it (and I didn't do anything, it just popped back by itself in half a second)

nopara73 commented:
This is great work! I barely have anything meaningful to report!

# Windows 10

No issues found.

# VirtualBox: Fedora (1 CPU)

## Issue 1 (not urgent)

It uses a different font, than Ubuntu. It may not need to be fixed at all, I couldn't see any places where it'd be an issue, but it's good to be aware of.

# VirtualBox: Ubuntu 16.04 (2 CPU)

No issues found.

# VirtualBox: OSX High Sierra (2 CPU)

## Issue 1 (not urgent)

The window is too large, I cannot place it to my OSX VM desktop:

![image](https://user-images.githubusercontent.com/9156103/46870921-48353a00-ce30-11e8-9c0f-2da4e1c94c7e.png)

There is no urgent need to fix this, since no sane person would use OSX in a VM, but it may comes up at other places, too. Anyway, I will open it as an issue after merging.

## Issue 2

After closing an app, an error comes up constantly.

![image](https://user-images.githubusercontent.com/9156103/46871034-93e7e380-ce30-11e8-948e-1751d1c9258f.png)

This error is not logged to the console.

Issue 3

ctrl+c/v vs cmd+c/v

So, I'm not quite sure what is the correct behavior, but it seems like other softwares on my OSX vm works with windows+c and windows+v, but Wasabi works with ctrl+c and ctrl+c (and all related stuff)  
I guess the windows button on my keyboard would be a cmd button on an OSX keyboard?  

@danwalmsley You have an OSX and you can see what's the expected behavior here.

Issue 4

Unable to close the application glitch:

![image](https://user-images.githubusercontent.com/9156103/46874037-7d458a80-ce38-11e8-83fa-4846a4f2f68c.png)

Zolgarr commented:
macOS High Sierra - Continuing the review

Issue 9 
a)When I clicked on settings, it was empty, and I couldn't close it either. 
b)Afterwards I dragged the arrow to the "maximize" icon, and it didn't react (the little sign didn't appear in it), but after a few seconds it was okay.

<img width="1440" alt="settings" src="https://user-images.githubusercontent.com/42039790/46872209-cfd07800-ce33-11e8-99f7-7783982e9888.png">

Issue 10
If I move my cursor on of the the "x" of settings/coinjoin/history/etc, there is no highlighting on the "x" itself. 

<img width="227" alt="highlighting the x" src="https://user-images.githubusercontent.com/42039790/46872476-85033000-ce34-11e8-934c-fe53979d260f.png">

Issue 11
If I go to Help--»Dev Tools--»Logical Tree or Visual Tree or Events: the font with black is not so visible.

<img width="478" alt="logical tree" src="https://user-images.githubusercontent.com/42039790/46872919-b8928a00-ce35-11e8-823d-18b4034a482e.png">

<img width="1440" alt="events" src="https://user-images.githubusercontent.com/42039790/46872932-bdefd480-ce35-11e8-87b3-398f18d44030.png">

lontivero commented:
The issue 2 happened to me too. 

issue 12: I open Wasabi, load a wallet and it starts downloading filters. I close the Wasabi  windows and it closes but the terminal console continues displaying the filters being downloaded:

```
2018-10-12 09:40:01 INFO: Starting WalletService...
2018-10-12 09:40:01 INFO: WalletService started.
2018-10-12 09:40:02 INFO IndexDownloader: Downloaded filters for blocks from 828576 to 829575.
2018-10-12 09:40:06 INFO CcjClientState: Round (9) added.
2018-10-12 09:40:06 INFO CcjClientState: Round (10) added.
2018-10-12 09:40:07 INFO IndexDownloader: Downloaded filters for blocks from 829576 to 830575.
2018-10-12 09:40:07 INFO CcjClient: CcjClient is successfully initialized.
2018-10-12 09:40:08 INFO IndexDownloader: Downloaded filters for blocks from 830576 to 831575.
2018-10-12 09:40:09 INFO IndexDownloader: Downloaded filters for blocks from 831576 to 832575.
2018-10-12 09:40:11 INFO IndexDownloader: Downloaded filters for blocks from 832576 to 833575.
2018-10-12 09:40:12 INFO IndexDownloader: Downloaded filters for blocks from 833576 to 834575.
2018-10-12 09:40:13 INFO IndexDownloader: Downloaded filters for blocks from 834576 to 835575.
2018-10-12 09:40:14 INFO IndexDownloader: Downloaded filters for blocks from 835576 to 836575.
2018-10-12 09:40:15 INFO IndexDownloader: Downloaded filters for blocks from 836576 to 837575.
2018-10-12 09:40:17 INFO IndexDownloader: Downloaded filters for blocks from 837576 to 838575.
2018-10-12 09:40:19 INFO IndexDownloader: Downloaded filters for blocks from 838576 to 839575.
2018-10-12 09:40:21 INFO IndexDownloader: Downloaded filters for blocks from 839576 to 840575.
2018-10-12 09:40:22 INFO IndexDownloader: Downloaded filters for blocks from 840576 to 841575.
2018-10-12 09:40:23 INFO IndexDownloader: Downloaded filters for blocks from 841576 to 842575.
2018-10-12 09:40:25 INFO IndexDownloader: Downloaded filters for blocks from 842576 to 843575.
2018-10-12 09:40:26 INFO IndexDownloader: Downloaded filters for blocks from 843576 to 844575.
2018-10-12 09:40:27 INFO IndexDownloader: Downloaded filters for blocks from 844576 to 845575.
```

danwalmsley commented:
@lontivero the cursor issue is fixed

danwalmsley commented:
@lontivero I can't tell if issue 12 is specific to OSX?

danwalmsley commented:
@nopara73 regarding CTRL + C, CMD + C

just waiting for this to be merged.. https://github.com/AvaloniaUI/Avalonia/pull/1964/files

lontivero commented:
This is the first time I see the issue nbr 12 that's why I think it could be Mac specific.

The problem with the icon that I reported is still in ca6d7b and the problem seems to be that there is a invisible horizontal line between the window header and the menu bar that when the mouse pass over it, it changes suggesting it can be drag to resize. Take a look at the image.

![image](https://user-images.githubusercontent.com/127973/46892476-63994880-ce22-11e8-96e2-137741d62fc9.png)

lontivero commented:
Mouse cursor issue was fixed. I was testing a previous commit, it seems.

danwalmsley commented:
@nopara73 
The remaining issues:
- are error message on close, (crashes in native part, will need more time to fix)
- key gestures CMD + C, etc this will need support in Avalonia and out of scope of this PR, its not a simple mod to do it properly.
- Render slowness, this will be fixed in a coming update to Avalonia.Native.

Iv addressed all the other issues that were possible to fix.
My suggestion is we should get this merged now, try and get more people testing it, knowing that there are the above known issues and hopefully we get more feedback and can track those issue separately.

Then we can next week ship an update to fix the rendering slowness, and the error message on app close.
